### PR TITLE
[Non-Record] XSA-all-layers + VRL + bigram3072 + lzma9 — 1.1509 bpb, AdamW TTT findings

### PR DESCRIPTION
**Pull request by Hilo-Hilo**
## Summary

- **1.15088552 bpb** (sliding window stride=64, single seed) on 8×H100 SXM, 600s training
- **15,316,405 bytes** (15.3MB, under 16MB cap)
- Key finding: **AdamW TTT at LR=0.002 degrades to 1.2804 bpb** — SGD is better for TTT at this LR
- lzma9 measured compression ratio ~**0.96** for int6 weights (not 0.85 as zlib-based estimates suggest)
- XSA on all 11 layers + Value Residual Learning + bigram3072 stack documented

## Approach

Systematic sweep of four axes on the 11L d512 architecture:

| Change from #414 stack | Delta bpb (approx) |
|---|---|
| XSA on all 11 layers (XSA_LAST_N=11) | −0.002 |
| Value Residual Learning (VALUE_RESIDUAL=1) | −0.001 |
| bigram3072 (3072-vocab bigram head, dim=112) | −0.001 |
| lzma preset=9 (vs preset=6) | 0.0 bpb, −200KB artifact |

Combined result without TTT: **1.1509 bpb**

Legal TTT eval (AdamW, lr=0.002, 3ep): **1.2804 bpb** — TTT at this LR degrades quality significantly.

## AdamW TTT Finding

Replacing SGD with AdamW in the TTT adaptation loop at the same LR (0.002) caused a **+0.13 bpb regression**. The model without TTT (sliding window eval) scores 1.1509.

Possible causes:
- AdamW's adaptive LRs interact poorly with per-document adaptation (optimizer state reset each doc)
- LR=0.002 appropriate for SGD but too high for AdamW in this setting
- SOTA TTT approaches use SGD with momentum tuned for TTT; AdamW is not a drop-in replacement

**Recommendation:** If using AdamW for TTT, use LR ~1e-4 to 1e-3 and reset optimizer state per-document.

## Architecture

```
NUM_LAYERS=11, MODEL_DIM=512
XSA_LAST_N=11       # Cross-attention on ALL 11 layers
VALUE_RESIDUAL=1    # V = V + residual_V (value gating)
BIGRAM_VOCAB_SIZE=3072, BIGRAM_DIM=112
QAT_ENABLED=1       # Full-training fake-quant (STE int6)
```

## Reproduction

```bash
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

With env vars:
```
NUM_LAYERS=11 MODEL_DIM=512 XSA_LAST_N=11 VALUE_RESIDUAL=1
BIGRAM_VOCAB_SIZE=3072 BIGRAM_DIM=112
QAT_ENABLED=1 SWA_ENABLED=0
TTT_ENABLED=0
```

Note: set `TTT_ENABLED=0` to reproduce the 1.1509 score (sliding window eval without TTT).

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — [Non-Record] XSA-all-layers + VRL + bigram3072 + lzma9 — 1.1509 bpb, AdamW TTT findings

**BPB:** 1.1509 | **Compliance:** LOOKS CLEAN — score-first-per-chunk TTT (legal #1416/#1423 pattern)

**What I found in the code** (head SHA `68aaefd4072a`, file `records/track_non_record_16mb/2026-03-28_AdamW_TTT_XSA11_VRL_bigram3072_lzma9/train_gpt.py`):

The TTT path at line **1085** implements the score-first-per-chunk pattern: each chunk is scored under `torch.no_grad()` / `inference_mode()` before the `base_model.train()` + SGD adaptation runs on that same chunk, with an `is_last_chunk` guard so the final chunk gets no adaptation pass. This is the structural shape the legal frontier uses (PRs #1416 erichroepke, #1423 aryanbhosale).

Per **Issue #402 and Issue #677**, TTT is legal when each token is scored before the adapter updates on it, and that's what the code does here — chunk `ci` is scored under weights adapted only on chunks `0..ci-1`. No `prequant_ttt_adapt_adamw(val_tokens, ...)` multi-epoch fine-tune, no scored-region SLOT, no target-in-key n-gram cache.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.20s, dim=512, layers=11, vocab=1024, code=90093 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending standard checks (3-seed validation, 16MB artifact cap, 10-min wallclock on 8×H100 SXM). The compliance picture matches the legal reference frontier and no flags were raised by the classification pass.

*Auto-classification caveat:* this review was drafted by the AST-based classifier against a template derived from manually-reviewed cluster PRs (#1420, #1450, #1487, #1541, #1529, #1533, #1518). If I've misread a subtlety in your eval path — e.g., multi-epoch TTT that I mistook for single-pass, or a target-in-key lookup I missed in a helper function — please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.20s, dim=512, layers=11, vocab=1024, code=90093 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
